controllers/component_pipelinerun_nudge_controller.go

```diff
@@ -0,0 +1,220 @@
+		log.Error(err, "failed to get component")
+		return ctrl.Result{}, err
+	}
+
+	if len(component.Spec.BuildNudgesRef) == 0 {
+		log.Info(fmt.Sprintf("component %s has no BuildNudgesRef set", component.Name))
+		return ctrl.Result{}, err
+	}
+
+	if pipelineRun.IsDone() || pipelineRun.DeletionTimestamp != nil {
+		if controllerutil.ContainsFinalizer(pipelineRun, NudgeFinalizer) || pipelineRun.Annotations == nil || pipelineRun.Annotations[NudgeProcessedAnnotationName] == "" {
+			//pipeline run is done and we have not cleared the finalizer yet
+			//we need to perform our nudge
+			return r.handleCompletedBuild(ctx, log, pipelineRun, component)
+		}
+	} else if !controllerutil.ContainsFinalizer(pipelineRun, NudgeFinalizer) {
+		//we add a finalizer to make sure we see the run before it is deleted
+		//as tekton results should aggressivly delete when pruning is enabled
+		controllerutil.AddFinalizer(pipelineRun, NudgeFinalizer)
+		err := r.Client.Update(ctx, pipelineRun)
+		if err != nil {
+			log.Error(err, "failed to add finalizer")
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ComponentNudgeReconciler) handleCompletedBuild(ctx context.Context, log logr.Logger, pipelineRun *tektonapi.PipelineRun, updatedComponent *applicationapiv1alpha1.Component) (ctrl.Result, error) {
+
+	//note that we do this eagerly, as we only realy have one chance to do the nudge
+	//we don't want to get stuck in a nudge loop, or fail to remove the finalizer
+	if pipelineRun.Annotations == nil {
+		pipelineRun.Annotations[NudgeProcessedAnnotationName] = "true"
+	}
+	controllerutil.RemoveFinalizer(pipelineRun, NudgeFinalizer)
+	err := r.Client.Update(ctx, pipelineRun)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	success := pipelineRun.Status.GetCondition(apis.ConditionSucceeded).IsTrue()
+	if !success {
+		log.Error(err, "Not performing nudge as pipeline failed")
+		return ctrl.Result{}, nil
+	}
+
+	components := applicationapiv1alpha1.ComponentList{}
+	err = r.Client.List(ctx, &components, client.InNamespace(pipelineRun.Namespace))
+	if err != nil {
+		log.Error(err, "failed to list components in namespace")
+		return ctrl.Result{}, err
+	}
+
+	image := ""
+	digest := ""
+	for _, r := range pipelineRun.Status.PipelineResults {
+		if r.Name == ImageDigestParam {
+			digest = r.Value.StringVal
+		} else if r.Name == ImageUrlParam {
+			image = r.Value.StringVal
+		}
+	}
+	if image == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageUrlParam), "no image url result")
+		return ctrl.Result{}, nil
+	}
+	if digest == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageDigestParam), "no image digest result")
+		return ctrl.Result{}, nil
+	}
+
+	for i := range components.Items {
+		comp := components.Items[i]
+		if comp.Spec.Application == updatedComponent.Spec.Application && slices.Contains(updatedComponent.Spec.BuildNudgesRef, comp.Name) {
+			//TODO: Do the nudge
```

Comment:
Nudge would be a complicated operation that takes some time and involves dealing with git provider API.
What if nudge for one of the components fails? What if it's transient/consistent error?

Reply:
This story was just about getting it to the point where we can print the message in the logs. I assume we will need a retry count mechanism with back off, we don't want to retry indefinitely, but we also don't want a single transient failure to stop the update.

Reply:
We will need to add an annotation with the state information. Also nudging should be idempotent, if you run the same nudge again it should figure out that there is an existing PR, and the update to the PR would result in the same commit.

Reply:
I have added support for this.

Reply:
Something to bear in mind with error handling too is that the actual updates will be async, and error handling will also need to be done through watching the job/pipeline that is doing the update.